Tests.java

```diff
@@ -0,0 +1,247 @@
+        int [] initial = {0,1,2};
+        int [] expected = initial;
+        assertEquals("", RearrangingCars.rearrangingSteps(initial, expected));
+    } 
+    
+    //Test when only one step is required
+    @Test
+    public void test_oneStepRearrangements() throws RearrangementExceptions {
+        int [] initial = {0,1,2};
+        int [] expected = {1,0,2};
+        ArrayList<Step> steps = new ArrayList<Step>();
+        steps.add(new Step(1,0));
+        assertEquals(steps.toString(), RearrangingCars.rearrangingSteps(initial, expected));
+    } 
+    
+    //Test when each car has to be shifted one position to its left: n-1 steps required
+    @Test
+    public void test_shiftLeftRearrangement() throws RearrangementExceptions {
+        int [] initial = {0,1,2,3};
+        int [] expected = {1,2,3,0};
+        ArrayList<Step> steps = new ArrayList<Step>();
+        steps.add(new Step(1,0));
+        steps.add(new Step(2,1));
+        steps.add(new Step(3,2));
+                
+        String resultSteps = RearrangingCars.rearrangingSteps(initial, expected);   
+        
+        assertEquals(steps.toString(), resultSteps.toString());
+    } 
+    
+    //Test when each car has to be shifted one position to its right (and rotated) : n-1 steps required
+    @Test
+    public void test_rotateShiftRightRearrangement() throws RearrangementExceptions {
+        int [] initial = {0,1,2,3};
+        int [] expected = {3,0,1,2};
+        ArrayList<Step> steps = new ArrayList<Step>();
+        steps.add(new Step(3,0));
+        steps.add(new Step(2,3));
+        steps.add(new Step(1,2));
+                
+        String resultSteps = RearrangingCars.rearrangingSteps(initial, expected);   
+        
+        assertEquals(steps.toString(), resultSteps.toString());
+    } 
+    
+    //Test when the initial and expected rearrangements have the empty slot at the same index
+    @Test
+    public void test_sameEmptySlotIndex() throws RearrangementExceptions {
+        int [] initial = {0,1,2};
+        int [] expected = {0,2,1};
+        ArrayList<Step> steps1 = new ArrayList<Step>();
+        steps1.add(new Step(1,0));
+        steps1.add(new Step(2,1));
+        steps1.add(new Step(0,2));
+        
+        ArrayList<Step> steps2 = new ArrayList<Step>();
+        steps2.add(new Step(2,0));
+        steps2.add(new Step(1,2));
+        steps2.add(new Step(0,1));
+        
+        HashSet<String> expectedStepsString = new HashSet<String>();
+        expectedStepsString.add(steps1.toString());
+        expectedStepsString.add(steps2.toString());
+        
+        String resultSteps = RearrangingCars.rearrangingSteps(initial, expected);
+        
+        assertTrue(expectedStepsString.toString().contains(resultSteps));
+    } 
+    
+    //Test when two pairs of slots need to be swapped: 6 steps required
+    @Test
+    public void test_twoswaps() throws RearrangementExceptions {

```

Comment:
I don't think that writing all combinations of possible solutions is a good approach to tests :) It is really hard to maintain it and after changing initial array a bit you have to rewrite all expected moves.